libs/net/dal/Services/FileReferenceService.cs

```diff
@@ -254,5 +258,155 @@ public async Task<IEnumerable<FileReference>> GetFiles(DateTime? updatedBefore =
+        var bucketName = Environment.GetEnvironmentVariable("S3_BUCKET_NAME") ?? throw new InvalidOperationException("S3_BUCKET_NAME environment variable is not set");
+        var serviceUrl = Environment.GetEnvironmentVariable("S3_SERVICE_URL") ?? throw new InvalidOperationException("S3_SERVICE_URL environment variable is not set");
+
+        var config = new AmazonS3Config
+        {
+            ServiceURL = serviceUrl,
+            ForcePathStyle = true
+        };
+
+        using var s3Client = new AmazonS3Client(accessKey, secretKey, config);
+
+        try
+        {
+            var request = new GetObjectRequest
+            {
+                BucketName = bucketName,
+                Key = s3Key
+            };
+
+            var response = await s3Client.GetObjectAsync(request);
+
+            if (response.HttpStatusCode == System.Net.HttpStatusCode.OK)
+            {
+                var memoryStream = new MemoryStream();
+                await response.ResponseStream.CopyToAsync(memoryStream);
+                memoryStream.Position = 0;
+                return memoryStream;
+            }
+            else
+            {
+                Logger.LogError("Failed to download file from S3: {S3Key}", s3Key);
+                throw new Exception($"Failed to download file from S3: {s3Key}");
+            }
+        }
+        catch (AmazonS3Exception ex)
+        {
+            Logger.LogError(ex, "Error retrieving file from S3: {S3Key}", s3Key);
+            throw;
+        }
+        catch (Exception ex)
+        {
+            Logger.LogError(ex, "Unexpected error when downloading file from S3: {S3Key}", s3Key);
+            throw;
+        }
+    }
+
+    public async Task<FileReference?> GetByS3PathAsync(string s3Path)
+    {
+        return await this.Context.FileReferences
+            .FirstOrDefaultAsync(fr => fr.S3Path == s3Path);
+    }
+
+    public async Task<FileReference?> GetByPathAsync(string path)
+    {
+        return await this.Context.FileReferences
+            .FirstOrDefaultAsync(fr => fr.Path == path);
+    }
+
+
+    public async Task<(Stream? Stream, string? FileName, string? ContentType)> GetFileStreamAsync(string path)
+    {
+        var fileReference = await GetByS3PathAsync(path);
+        if (fileReference == null)
+        {
+            fileReference = await GetByPathAsync(path);
+        }
+        if (fileReference == null)
+        {
+            Logger.LogInformation("File reference not found for path: {Path}", path);
+            return (null, null, null);
+        }
+
+        if (fileReference.IsSyncedToS3)
+        {
+            try
+            {
+                var stream = await DownloadFromS3Async(path);
+                return (stream, fileReference.FileName, fileReference.ContentType);
+            }
+            catch (Exception ex)
+            {
+                Logger.LogError(ex, "Error on stream file from S3: {Path}", path);
```

Comment:
Why does it try to get the file at this location if the file doesn't exist? Logging an error indicates an issue.